### DISCUSSION TITLE
Trouble Reading Parquet Files from Azure Blob Storage with DuckDB

### DISCUSSION THREAD

**Discussion by radix-naimsulejmani** [duckdb/duckdb#8574] [General]
**Environment:**

  Data Source: Multiple
  Database: DuckDB
  Service: Azure Function
  Language: Typescript

**Description**:
I've been attempting to migrate data from various sources into DuckDB and subsequently parse it into a Parquet file format. When saving the data directly, everything works seamlessly. However, I've been encountering challenges when reading the saved data. As of now, I'm required to download the file from Azure Blob Storage to local storage first before querying it.

I've used the Azurite library for testing. To ensure this wasn't an isolated issue, I also tested with a real connection to Azure Blob Storage. Both methods result in a 403 error.

```sql
INSTALL httpfs;
LOAD httpfs;
SET s3_endpoint='127.0.0.1:10000';
SET s3_access_key_id='devstoreaccount1';
SET s3_secret_access_key='[REDACTED]';  
SET s3_url_style='path';
SET s3_use_ssl='false';

SELECT * FROM parquet_scan('s3://devstoreaccount1/parquets/geolocation.parquet');
```
I'd appreciate any insights or suggestions to address this issue.

Error Message:
```vbnet
HTTP Error: HTTP GET error on 'http://127.0.0.1:10000/devstoreaccount1/parquets/geolocation.parquet' (HTTP 403)
Code Snippet:
```

**Comment by Mause** [comment]
Hey there, this is expected, as Azure Blob Storage doesn't support the S3 protocol.

Your options are:
 * use the S3Proxy program Azure provides: https://devblogs.microsoft.com/cse/2016/05/22/access-azure-blob-storage-from-your-apps-using-s3-api/
 * Use the support provided by our fsspec integration
 * Use the (for now experimental, and only available in our dev builds) `azure` extension: https://github.com/duckdblabs/duckdb_azure

**Reply by radix-naimsulejmani** [reply]
Thanks @Mause , 
I will try that, especially the duckdb_azure, as I see it will be released in version 0.9 ;)

**Reply by radix-naimsulejmani** [reply]
Hello @Mause,

I wanted to share that I've found an alternative approach for reading Azure Blob Storage Parquet files directly. The method involves generating a short-lived SAS (Shared Access Signature) Token with read permissions specifically for the file or blob container in question.

Here's a SQL example demonstrating how it works:
```SQL
INSTALL httpfs;
LOAD httpfs;

SELECT * FROM "http://127.0.0.1:10000/devstoreaccount1/parquets/persons.parquet?sv=2023-01-03&se=2023-08-23T19%3A49%3A30Z&sr=b&sp=rwde&sig=50%2F%2FoNGeyOu6oA1hQY5P6hCKkQYtkZaxrwokyYr703U%3D";
```
The SAS Token can be generated through various programming languages such as Python, JavaScript, TypeScript, C#, or Java. Additionally, you can also generate it via the Azure portal.


However, it's important to note that even though the SAS Token grants write access, it can't be used to directly upload or write files to Azure Blob Storage using the same method.

```SQL
COPY (SELECT id,name FROM persons) TO "http://127.0.0.1:10000/devstoreaccount1/parquets/persons.parquet?sv=2023-01-03&se=2023-08-23T19%3A49%3A30Z&sr=b&sp=rwde&sig=50%2F%2FoNGeyOu6oA1hQY5P6hCKkQYtkZaxrwokyYr703U%3D" (FORMAT PARQUET, COMPRESSION ZSTD)
;
```
